### PR TITLE
Support private acl for objects

### PR DESCRIPTION
**Pull request by lelikg**
allow to set private ACL on objects

### CODE REVIEW

File: fakestorage/response.go [hunk 1]
```diff
@@ -4,7 +4,14 @@
 
 package fakestorage
 
-import "sort"
+import (
+	"net/http"
+	"sort"
+
+	cloudstorage "cloud.google.com/go/storage"
```
Comment (fsouza) [inline fakestorage/response.go]:
```suggestion
	"cloud.google.com/go/storage"
```

File: fakestorage/object.go [hunk 2]
```diff
@@ -7,11 +7,14 @@ package fakestorage
...
 	"encoding/json"
 	"fmt"
+	"io/ioutil"
 	"net/http"
 	"sort"
 	"strconv"
 	"strings"
 
+	"cloud.google.com/go/storage"
+
```
Comment (fsouza) [inline fakestorage/object.go]: Please delete this line.

File: fakestorage/object.go [hunk 3]
```diff
@@ -174,6 +180,48 @@ func (s *Server) deleteObject(w http.ResponseWriter, r *http.Request) {
...
+	bodyMap := make(map[string]string)
+	defer r.Body.Close()
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "bad request", http.StatusBadRequest)
+		return
+	}
+
+	json.Unmarshal(body, &bodyMap)
```
Comment (fsouza) [inline fakestorage/object.go]: Please check the error here. Also, where is bodyMap being used?

File: fakestorage/object_test.go [hunk 4]
```diff
@@ -604,3 +604,42 @@ func TestServerClientObjectDeleteErrors(t *testing.T) {
...
+
+	runServersTest(t, objs, func(t *testing.T, server *Server) {
+		t.Run("public to private", func(t *testing.T) {
+			ctx := context.Background()
+			objHandle := server.Client().Bucket("some-bucket").Object("img/public-to-private.jpg")
+
+			err := objHandle.ACL().Set(ctx, storage.AllAuthenticatedUsers, storage.RoleReader)
+			if err != nil {
+				t.Errorf("unexpected error while setting acl %+v", err)
+				return
```
Comment (fsouza) [inline fakestorage/object_test.go]: use ``t.Fatalf`` instead of ``t.Errorf`` + ``return``.

File: fakestorage/object_test.go [hunk 5]
```diff
@@ -604,3 +604,42 @@ func TestServerClientObjectDeleteErrors(t *testing.T) {
...
+			}
+
+			rules, err := objHandle.ACL().List(ctx)
+			if err != nil {
+				t.Errorf("unexpected error while getting acl %+v", err)
+				return
```
Comment (fsouza) [inline fakestorage/object_test.go]: same as above

File: fakestorage/object_test.go [hunk 6]
```diff
@@ -604,3 +604,42 @@ func TestServerClientObjectDeleteErrors(t *testing.T) {
...
+			}
+
+			if len(rules) != 1 {
+				t.Error("acl has no rules")
+				return
```
Comment (fsouza) [inline fakestorage/object_test.go]: ditto

File: fakestorage/object_test.go [hunk 7]
```diff
@@ -604,3 +604,42 @@ func TestServerClientObjectDeleteErrors(t *testing.T) {
...
+			}
+			if rules[0].Entity != storage.AllAuthenticatedUsers {
+				t.Error("acl entity not set to AllAuthenticatedUsers")
+				return
```
Comment (fsouza) [inline fakestorage/object_test.go]: ditto

File: fakestorage/object_test.go [hunk 8]
```diff
@@ -604,3 +604,42 @@ func TestServerClientObjectDeleteErrors(t *testing.T) {
...
+			}
+
+			if rules[0].Role != storage.RoleReader {
+				t.Error("acl role not set to RoleReader")
+				return
```
Comment (fsouza) [inline fakestorage/object_test.go]: ditto

File: fakestorage/response.go (hunk 1 above)
Comment (lelikg) [reply fakestorage/response.go]: this conflicts with "google.golang.org/api/storage/v1"

File: fakestorage/object.go (hunk 2 above)
Comment (lelikg) [reply fakestorage/object.go]: I am using storage.ACLEntity in line 31

File: fakestorage/object.go (hunk 3 above)
Comment (lelikg) [reply fakestorage/object.go]: removed

File: fakestorage/object_test.go (hunk 4 above)
Comment (lelikg) [reply fakestorage/object_test.go]: done

File: fakestorage/object_test.go (hunk 6 above)
Comment (lelikg) [reply fakestorage/object_test.go]: done

File: fakestorage/object_test.go (hunk 7 above)
Comment (lelikg) [reply fakestorage/object_test.go]: done

File: fakestorage/object_test.go (hunk 5 above)
Comment (lelikg) [reply fakestorage/object_test.go]: done

File: fakestorage/object_test.go (hunk 8 above)
Comment (lelikg) [reply fakestorage/object_test.go]: done

**Comment by lelikg** [comment]
> Looks mostly good! Wonder if we should add support for ACLs in other endpoints too (like object creation)?

IMHO it's not possible to set ACL on object creation, you can only do it after the object is already created.

File: fakestorage/object.go (hunk 2 above)
Comment (fsouza) [reply fakestorage/object.go]: yeah I meant the blank line 🙈

Comment (lelikg) [reply fakestorage/object.go]: done

**Comment by fsouza** [comment]
> > Looks mostly good! Wonder if we should add support for ACLs in other endpoints too (like object creation)?
> 
> IMHO it's not possible to set ACL on object creation, you can only do it after the object is already created.

What does this code do? Can we add a test to make sure it behaves as expected?

```go
package main

import (
	"context"

	"cloud.google.com/go/storage"
)

func main() {
	client, err := storage.NewClient(context.Background())
	if err != nil {
		panic(err)
	}
	obj := client.Bucket("some-bucket").Object("some/new-object.txt")
	writer := obj.NewWriter(context.Background())
	rule := storage.ACLRule{
		Entity: storage.AllUsers,
		Role:   storage.RoleReader,
	}
	writer.ObjectAttrs.ACL = append(writer.ObjectAttrs.ACL, rule)
	writer.Write([]byte("hi there"))
	err = writer.Close()
	if err != nil {
		panic(err)
	}
}
```

(I haven't tried running it, I can investigate later though)

**Comment by lelikg** [comment]
> ```go
> storage.NewClient
> ```

Unfortunately the ACL is not written to the object when doing it this way

**Comment by fsouza** [comment]
@lelikg sounds good. It looks like it can take PredefinedACL, but that's a different thing.

Thanks for contributing!

**Comment by lelikg** [comment]
thanks for merging, is it possible to trigger a release?

**Comment by fsouza** [comment]
Done, just pushed v1.13.0.

**Pull request merged**